### PR TITLE
Add Historical Staking Balances & USD Conversion Information

### PR DESCRIPTION
**Pull request by deangalvin-cb**
### What changed? Why?

This adds historical staking balance information as well as USD conversion information (date & price)

#### Qualified Impact
This adds `usd_conversion_price` `usd_conversion_time` and `Coinbase::StakingBalance` methods/objects. To allow developers to provide more accurate staking reward information to their users.

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

File: lib/coinbase/client/api/external_addresses_api.rb [hunk 1]
```diff
@@ -94,6 +94,91 @@ def get_external_address_balance_with_http_info(network_id, address_id, asset_id
...
+      # verify the required parameter 'asset_id' is set
+      if @api_client.config.client_side_validation && asset_id.nil?
+        fail ArgumentError, "Missing the required parameter 'asset_id' when calling ExternalAddressesApi.list_address_historical_balance"
+      end
+      if @api_client.config.client_side_validation && !opts[:'page'].nil? && opts[:'page'].to_s.length > 5000
+        fail ArgumentError, 'invalid value for "opts[:"page"]" when calling ExternalAddressesApi.list_address_historical_balance, the character length must be smaller than or equal to 5000.'
+      end
+
+      # resource path
+      local_var_path = '/v1/networks/{network_id}/addresses/{address_id}/balance_history/{asset_id}'.sub('{' + 'network_id' + '}', CGI.escape(network_id.to_s)).sub('{' + 'address_id' + '}', CGI.escape(address_id.to_s)).sub('{' + 'asset_id' + '}', CGI.escape(asset_id.to_s))
```
Comment (marchsu) [inline lib/coinbase/client/api/external_addresses_api.rb:141]: this is not for staking balance but general account balance, I assume we are adding the support for checking general account balance as well, but did not see [the staking balance equivalent ](https://github.cbhq.net/c3/cdp-service/blob/4b2ae9dabe6de4c494c1fdbd072bffe0deb8a02f/openapi/cdp-service.yml#L1794)
`/v1/networks/{network_id}/addresses/{address_id}/stake/balances`

File: lib/coinbase/client/api/addresses_api.rb [hunk 2]
```diff
@@ -6,7 +6,7 @@
 The version of the OpenAPI document: 0.0.1-alpha
 Contact: yuga.cohler@coinbase.com
 Generated by: https://openapi-generator.tech
-Generator version: 7.7.0
+Generator version: 7.6.0
```
Comment (marchsu) [inline lib/coinbase/client/api/addresses_api.rb:9]: seeing a lot of this Generator change, is it intended?

File: lib/coinbase/staking_balance.rb [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+require 'date'
+
+module Coinbase
+  # A representation of a staking balance on a network for a given asset.
+  class StakingBalance
+    # Returns a list of StakingBalance for the provided network, asset, and addresses.
+    # @param network_id [Symbol] The network ID
+    # @param asset_id [Symbol] The asset ID
+    # @param address_ids [Array<String>] The address IDs
```
Comment (marchsu) [inline lib/coinbase/staking_balance.rb:11]: unlike staking rewards, staking balance query accepts only 1 address at a time, our rewards client will report this error when querying with `address=a or address=b`
```
{
  "code": "bad_staking_request",
  "message": "Filter validation failed. invalid filter provided: filter only accepts additive expressions (i.e. using AND). using OR, NOT, or other expressions isn't supported. please use only AND expressions"
}
```

File: lib/coinbase/staking_balance.rb [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    def inspect
+      to_s
+    end
+
+    def self.stake_api
+      Coinbase::Client::StakeApi.new(Coinbase.configuration.api_client)
+    end
+
+    def self.list_page(network_id, asset_id, address_ids, start_time, end_time, page)
+      stake_api.fetch_historical_staking_balances(
```
Comment (marchsu) [inline lib/coinbase/staking_balance.rb:76]: can't find the implementation of  `fetch_historical_staking_balances`, or I missed something? not familiar with ruby

File: lib/coinbase/client/api/addresses_api.rb (hunk 2 above)
Comment (deangalvin-cb) [reply lib/coinbase/client/api/addresses_api.rb:9]: Yes, these are automated

File: lib/coinbase/client/api/external_addresses_api.rb (hunk 1 above)
Comment (deangalvin-cb) [reply lib/coinbase/client/api/external_addresses_api.rb:141]: This is automated

File: lib/coinbase/staking_balance.rb (hunk 4 above)
Comment (deangalvin-cb) [reply lib/coinbase/staking_balance.rb:76]: This is from the generated client.

**Review by alex-stone — APPROVED** [review]

**Pull request merged**